src/backend/distributed/connection/placement_connection.c

```diff
@@ -229,272 +222,439 @@ GetPlacementConnection(uint32 flags, ShardPlacement *placement, const char *user
+				 * the same connection, or errored out). Update the connection reference
+				 * to point to the connection used for writing. We don't need to remember
+				 * the existing connection since we won't be able to reuse it for
+				 * accessing the placement. However, we do register that it exists in
+				 * hasSecondaryConnections.
+				 */
+				placementConnection->connection = chosenConnection;
+				placementConnection->userName = MemoryContextStrdup(TopTransactionContext,
+																	userName);
+
+				AssertArg(!placementConnection->hadDDL);
+				AssertArg(!placementConnection->hadDML);
+			}
+
+			/*
+			 * There are now multiple connections that read from the placement
+			 * and DDL commands are forbidden.
+			 */
+			placementEntry->hasSecondaryConnections = true;
+
+			if (placementEntry->colocatedEntry != NULL)
+			{
+				/* we also remember this for co-located placements */
+				placementEntry->colocatedEntry->hasSecondaryConnections = true;
+			}
 		}
 
 		/*
-		 * If not using the connection, verify that FOR_DML/DDL flags are
-		 * compatible.
+		 * Remember that we used the current connection for writes.
 		 */
-		if (flags & FOR_DDL)
+		if (accessType == PLACEMENT_ACCESS_DDL)
 		{
-			/* would deadlock otherwise */
-			ereport(ERROR,
-					(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
-					 errmsg("cannot establish new placement connection for DDL when "
-							"other connections exist for colocated placements")));
+			placementConnection->hadDDL = true;
 		}
-		else if (flags & FOR_DML)
+
+		if (accessType == PLACEMENT_ACCESS_DML)
 		{
+			placementConnection->hadDML = true;
+		}
+	}
+
+	if (freeUserName)
+	{
+		pfree(freeUserName);
+	}
+
+	return chosenConnection;
+}
+
+
+/*
+ * FindOrCreatePlacementEntry finds a placement entry in either the
+ * placement->connection hash or the co-located placements->connection hash,
+ * or adds a new entry if the placement has not yet been accessed in the
+ * current transaction.
+ */
+static ConnectionPlacementHashEntry *
+FindOrCreatePlacementEntry(ShardPlacement *placement)
+{
+	ConnectionPlacementHashKey key;
+	ConnectionPlacementHashEntry *placementEntry = NULL;
+	bool found = false;
+
+	key.placementId = placement->placementId;
+
+	placementEntry = hash_search(ConnectionPlacementHash, &key, HASH_ENTER, &found);
+	if (!found)
+	{
+		/* no connection has been chosen for this placement */
+		placementEntry->failed = false;
+		placementEntry->primaryConnection = NULL;
+		placementEntry->hasSecondaryConnections = false;
+		placementEntry->colocatedEntry = NULL;
+
+		if (placement->partitionMethod == DISTRIBUTE_BY_HASH ||
+			placement->partitionMethod == DISTRIBUTE_BY_NONE)
+		{
+			ColocatedPlacementsHashKey key;
+			ColocatedPlacementsHashEntry *colocatedEntry = NULL;
+
+			strcpy(key.nodeName, placement->nodeName);
+			key.nodePort = placement->nodePort;
+			key.colocationGroupId = placement->colocationGroupId;
+			key.representativeValue = placement->representativeValue;
```

Comment:
I guess mathematically it doesn't matter much. `int` would seem a bit easier for debugging purposes as it matches pg_dist_shard, while `uint` is a bit easier to reason about.